// @cyborgtests/test/dist/index.mjs
// src/test.ts
import {
  expect,
  test as pwTest
} from "@playwright/test";
import { chromium } from "playwright";

// src/config/index.ts
var defaultConfig = {
  analyticsEnabled: true,
  uiPort: 3009
};
var userConfig = {};
var config = {
  setConfig: (config2) => {
    userConfig = { ...userConfig, ...config2 };
  },
  get analyticsEnabled() {
    return userConfig.analyticsEnabled ?? defaultConfig.analyticsEnabled;
  },
  get uiPort() {
    return userConfig.uiPort ?? defaultConfig.uiPort;
  }
};

// src/utils/server.ts
async function startServer(port) {
  const http = await import("http");
  const fs = await import("fs");
  const path = await import("path");
  const appBuildPath = path.resolve(process.cwd(), "node_modules/@cyborgtests/test/control-panel-build");
  const server = http.createServer((req, res) => {
    const url = req.url || "/";
    const filePath = path.join(appBuildPath, url === "/" ? "index.html" : url);
    fs.access(filePath, fs.constants.F_OK, (err) => {
      if (err) {
        res.writeHead(404);
        res.end("Not found");
        return;
      }
      const ext = path.extname(filePath).toLowerCase();
      const contentType = {
        ".html": "text/html",
        ".js": "text/javascript",
        ".png": "image/png",
        ".jpg": "image/jpeg",
        ".jpeg": "image/jpeg",
        ".svg": "image/svg+xml",
        ".ico": "image/x-icon"
      }[ext] || "application/octet-stream";
      fs.readFile(filePath, (err2, data) => {
        if (err2) {
          res.writeHead(500);
          res.end("Error loading file");
          return;
        }
        if (ext === ".html") {
          const html = data.toString();
          const configScript = `<script>window.ENABLE_ANALYTICS = ${config.analyticsEnabled};</script>`;
          const modifiedHtml = html.replace("</head>", `${configScript}</head>`);
          res.writeHead(200, { "Content-Type": contentType });
          res.end(modifiedHtml);
        } else {
          res.writeHead(200, { "Content-Type": contentType });
          res.end(data);
        }
      });
    });
  });
  let currentPort = port;
  const maxPortAttempts = 10;
  for (let i = 0; i < maxPortAttempts; i++) {
    try {
      await new Promise((resolve, reject) => {
        server.once("error", (err) => {
          if (err.code === "EADDRINUSE") {
            currentPort = port + i + 1;
            reject(err);
          } else {
            reject(err);
          }
        });
        server.listen(currentPort, () => {
          config.setConfig({
            uiPort: currentPort
          });
          resolve();
        });
      });
      break;
    } catch (err) {
      if (i === maxPortAttempts - 1) {
        throw new Error(`Failed to start server after ${maxPortAttempts} attempts`);
      }
      continue;
    }
  }
  await waitForServer(currentPort, http);
  const serverProcess = {
    kill: () => {
      server.close();
    }
  };
  return serverProcess;
}
async function waitForServer(port, http, maxAttempts = 10) {
  for (let i = 0; i < maxAttempts; i++) {
    try {
      await new Promise((resolve, reject) => {
        http.get(`http://localhost:${port}`, (res) => {
          if (res.statusCode === 200) {
            resolve();
          } else {
            reject(new Error(`Server returned ${res.statusCode}`));
          }
        }).on("error", reject);
      });
      return;
    } catch (err) {
      if (i === maxAttempts - 1) throw err;
      await new Promise((resolve) => setTimeout(resolve, 1e3));
    }
  }
}

// src/utils/openInDefaultBrowser.ts
async function openInDefaultBrowser(url) {
  const { exec } = await import("child_process");
  const os = await import("os");
  const currentPlatform = os.platform();
  const commands = {
    darwin: `open "${url}"`,
    win32: `start "" "${url}"`,
    linux: `xdg-open "${url}"`
  };
  const command = commands[currentPlatform];
  if (command) exec(command);
}
var openInDefaultBrowser_default = openInDefaultBrowser;

// src/utils/jira/config.ts
function getJiraConfig() {
  const baseUrl = process.env.JIRA_BASE_URL;
  const email = process.env.JIRA_EMAIL;
  const apiToken = process.env.JIRA_API_TOKEN;
  const projectKey = process.env.JIRA_PROJECT_KEY;
  if (!baseUrl || !email || !apiToken) {
    return null;
  }
  return {
    baseUrl,
    email,
    apiToken,
    projectKey
  };
}
function checkJiraConfig() {
  const config2 = getJiraConfig();
  return config2 !== null;
}
function getJiraProjectKey() {
  return process.env.JIRA_PROJECT_KEY || "";
}
function createAuthHeader(email, apiToken) {
  return Buffer.from(`${email}:${apiToken}`).toString("base64");
}

// src/utils/jira/issueTypes.ts
async function getJiraIssueTypes(projectKey) {
  const config2 = getJiraConfig();
  if (!config2 || !projectKey) {
    return [];
  }
  try {
    const auth = createAuthHeader(config2.email, config2.apiToken);
    const response = await fetch(
      `${config2.baseUrl}/rest/api/3/issue/createmeta?projectKeys=${projectKey}&expand=projects.issuetypes`,
      {
        method: "GET",
        headers: {
          "Authorization": `Basic ${auth}`,
          "Accept": "application/json"
        }
      }
    );
    if (!response.ok) {
      console.error("Failed to fetch Jira issue types:", response.status);
      return [];
    }
    const data = await response.json();
    const project = data.projects?.[0];
    if (!project) return [];
    return project.issuetypes.map((type) => ({
      id: type.id,
      name: type.name,
      description: type.description
    }));
  } catch (error) {
    console.error("Error fetching Jira issue types:", error);
    return [];
  }
}

// src/utils/jira/createTicket.ts
function convertToADF(description) {
  return {
    type: "doc",
    version: 1,
    content: description.split("\n").map((line) => ({
      type: "paragraph",
      content: line.trim() ? [{
        type: "text",
        text: line
      }] : []
    }))
  };
}
async function createJiraTicket(ticketInfo) {
  const config2 = getJiraConfig();
  if (!config2) {
    return { success: false, error: "Jira not configured" };
  }
  try {
    const auth = createAuthHeader(config2.email, config2.apiToken);
    const descriptionADF = convertToADF(ticketInfo.description);
    const issueData = {
      fields: {
        project: {
          key: ticketInfo.projectKey
        },
        summary: ticketInfo.summary,
        description: descriptionADF,
        issuetype: {
          name: ticketInfo.issueType
        }
      }
    };
    const response = await fetch(`${config2.baseUrl}/rest/api/3/issue`, {
      method: "POST",
      headers: {
        "Authorization": `Basic ${auth}`,
        "Accept": "application/json",
        "Content-Type": "application/json"
      },
      body: JSON.stringify(issueData)
    });
    if (!response.ok) {
      const errorData = await response.json();
      console.error("Jira API error:", errorData);
      return {
        success: false,
        error: errorData.errorMessages?.join(", ") || errorData.errors ? JSON.stringify(errorData.errors) : `HTTP error! status: ${response.status}`
      };
    }
    const result = await response.json();
    return { success: true, issueKey: result.key, issueId: result.id };
  } catch (error) {
    return {
      success: false,
      error: error instanceof Error ? error.message : "Unknown error"
    };
  }
}

// src/test.ts
var getFile = async () => {
  const fs = await import("fs/promises");
  const path = await import("path");
  const testBuilderAppPath = path.resolve(process.cwd(), "node_modules/@cyborgtests/test/test-builder-build");
  const html = await fs.readFile(path.join(testBuilderAppPath, "index.html"), "utf-8");
  const script = [...html.matchAll(/<script[^>]*>([\s\S]*?)<\/script>/gi)].map((match) => match[1]).join("\n");
  const styles = [...html.matchAll(/<style[^>]*>([\s\S]*?)<\/style>/gi)].map((match) => match[1]).join("\n");
  return { script, styles };
};
var TestFailedError = class extends Error {
  constructor(message) {
    super(message);
    this.name = "TestFailedError";
  }
};
var test = pwTest.extend({
  testControl: async ({ page, context, browser }, use, testInfo) => {
    let tcBrowser = null;
    let tcPage = null;
    let server = null;
    const getTestControl = async () => {
      if (!tcBrowser) {
        tcBrowser = await chromium.launch({
          headless: false
        });
        server = await startServer(config.uiPort);
        tcPage = await tcBrowser.newPage({
          viewport: { width: 500, height: 750 }
        });
        const { script, styles } = await getFile();
        await page.addInitScript(({ script: scriptContent, styles: stylesContent }) => {
          document.addEventListener("DOMContentLoaded", () => {
            const rootDiv = document.createElement("div");
            rootDiv.id = "cyborg-app";
            document.body.appendChild(rootDiv);
            const script2 = document.createElement("script");
            script2.type = "module";
            script2.textContent = scriptContent;
            document.head.appendChild(script2);
            const style = document.createElement("style");
            style.setAttribute("rel", "stylesheet");
            style.textContent = stylesContent;
            document.head.appendChild(style);
          });
        }, { script, styles });
        await tcPage.goto(`http://localhost:${config.uiPort}`);
        await tcPage.exposeFunction("openInMainBrowser", (link) => {
          openInDefaultBrowser_default(link);
        });
        await tcPage.evaluate(() => {
          window.testUtils.openInMainBrowser = window.openInMainBrowser;
        });
        await tcPage.exposeFunction("getTestInfo", () => {
          return {
            testId: testInfo.testId,
            attachments: testInfo.attachments,
            annotations: testInfo.annotations,
            project: testInfo.project,
            config: testInfo.config,
            title: testInfo.title,
            titlePath: testInfo.titlePath,
            file: testInfo.file,
            tags: testInfo.tags
          };
        });
        await tcPage.exposeFunction("skipTest", () => {
          if (testInfo.skip) {
            testInfo.skip(true);
          }
        });
        await tcPage.exposeFunction("checkJiraConfig", () => {
          return checkJiraConfig();
        });
        await tcPage.exposeFunction("getJiraProjectKey", () => {
          return getJiraProjectKey();
        });
        await tcPage.exposeFunction("getJiraIssueTypes", async (projectKey) => {
          return await getJiraIssueTypes(projectKey);
        });
        await tcPage.exposeFunction("createJiraTicket", async (ticketInfo) => {
          return await createJiraTicket(ticketInfo);
        });
        await tcPage.bringToFront();
      }
      return {
        browser: tcBrowser,
        context: tcPage.context(),
        page: tcPage
      };
    };
    const testControlObj = {
      get page() {
        throw new Error("testControl.page is not available. Use manualStep() to initialize the control panel.");
      },
      get browser() {
        throw new Error("testControl.browser is not available. Use manualStep() to initialize the control panel.");
      },
      get context() {
        throw new Error("testControl.context is not available. Use manualStep() to initialize the control panel.");
      },
      _getTestControl: getTestControl,
      _initialized: false
    };
    await use(testControlObj);
    if (tcPage) {
      await tcPage.close();
    }
    if (tcBrowser) {
      await tcBrowser.close();
    }
    if (server) {
      server.kill();
    }
  },
  manualStep: async ({ testControl, page, browser, context }, use) => {
    let currentResumeResolver = null;
    let resumeFunctionExposed = false;
    const manualStep = async (stepName, params = {}) => {
      test.setTimeout(0);
      return await test.step(
        `\u270B [MANUAL] ${stepName}`,
        async () => {
          const tc = await testControl._getTestControl();
          testControl._initialized = true;
          await tc.page.evaluate(
            ({ stepName: stepName2, params: params2 }) => {
              window.testUtils?.addStep(stepName2, params2);
            },
            { stepName, params }
          );
          const resumePromise = new Promise((resolve) => {
            currentResumeResolver = resolve;
          });
          if (!resumeFunctionExposed) {
            await tc.page.exposeFunction("resumeTest", () => {
              if (currentResumeResolver) {
                currentResumeResolver();
                currentResumeResolver = null;
              }
            });
            resumeFunctionExposed = true;
          }
          await tc.page.evaluate(() => {
            if (window.testUtils) {
              window.testUtils.resumeTest = window.resumeTest;
            }
          });
          await resumePromise;
          const hasFailed = await tc.page.evaluate(() => {
            if (window.testUtils.hasFailed) {
              delete window.testUtils.hasFailed;
              return true;
            }
            return false;
          });
          if (hasFailed) {
            const reason = await tc.page.evaluate(() => {
              const reason2 = window.testUtils?.failedReason || "";
              delete window.testUtils.failedReason;
              return reason2;
            });
            const errorMessage = `${stepName}${reason ? ` - ${reason}` : ""}`;
            throw new TestFailedError(errorMessage);
          }
        },
        { box: true, timeout: 0 }
      );
    };
    manualStep.soft = async (stepName, params = {}) => await test.step(
      `\u270B [MANUAL][SOFT] ${stepName}`,
      async () => {
        try {
          await manualStep(stepName, { isSoft: true, ...params });
        } catch (err) {
          test.info().annotations.push({
            type: "softFail",
            description: `Soft fail in manual step: ${err.message}`
          });
          await expect.soft(false, `Soft fail in manual step: ${err.message}`).toBeTruthy();
          console.warn(`Soft fail in manual step: ${stepName}`, err);
        }
      },
      { box: true, timeout: 0 }
    );
    await use(manualStep);
    try {
      if (currentResumeResolver) {
        currentResumeResolver();
        currentResumeResolver = null;
      }
    } catch (err) {
    }
  }
});
var test_default = test;

// src/index.ts
var index_default = test_default;
export {
  config,
  index_default as default
};
