// @cyborgtests/reporter-playwright-reports-server/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const promises_1 = __importDefault(require("fs/promises"));
const path_1 = __importDefault(require("path"));
const test_1 = require("@playwright/test");
const child_process_1 = require("child_process");
// Function to get user information
const getUsername = () => {
    let username = process.env.QA_USERNAME || '';
    if (username) {
        return username;
    }
    try {
        const gitUser = (0, child_process_1.execSync)('git config user.name', { encoding: 'utf8' }).trim();
        if (gitUser) {
            return gitUser;
        }
    }
    catch (error) {
        // Git config not available, continue with system user
    }
    return '';
};
const DEFAULT_OPTIONS = {
    enabled: true,
    resultDetails: {},
    triggerReportGeneration: true,
    requestTimeout: 60000,
};
class ReporterPlaywrightReportsServer {
    constructor(options) {
        this.rpOptions = { ...DEFAULT_OPTIONS, ...options };
        if (this.rpOptions.enabled === false) {
            return;
        }
        if (!this.rpOptions.reportPath) {
            throw new Error('[ReporterPlaywrightReportsServer] reportPath is required, cannot run without it');
        }
        this.blobPath = path_1.default.join(process.cwd(), this.rpOptions.reportPath);
        this.blobName = path_1.default.basename(this.blobPath);
        if (!this.rpOptions.url) {
            throw new Error('[ReporterPlaywrightReportsServer] url is required, cannot run without it');
        }
    }
    onBegin(config /*suite: Suite*/) {
        if (this.rpOptions.enabled === false) {
            return;
        }
        this.pwConfig = config;
    }
    async onEnd( /*result: FullResult*/) {
        if (this.rpOptions.enabled === false) {
            return;
        }
        let buffer;
        try {
            buffer = await promises_1.default.readFile(this.blobPath);
        }
        catch (err) {
            console.error(err);
            throw new Error('[ReporterPlaywrightReportsServer] Blob file not found or cannot be loaded. Results cannot be uploaded');
        }
        const ctx = await test_1.request.newContext({
            timeout: this.rpOptions.requestTimeout,
            extraHTTPHeaders: this.rpOptions.token !== undefined
                ? {
                    Authorization: this.rpOptions.token,
                }
                : {},
        });
        // Replace undefined values with empty strings
        const clearedResDetails = Object.fromEntries(Object.entries(this.rpOptions.resultDetails).map(([key, value]) => [key, value ?? '']));
        // Add username to result details if not already provided
        if (!clearedResDetails.username) {
            const username = getUsername();
            if (username)
                clearedResDetails.username = username;
        }
        const version = this.pwConfig.version ?? '';
        const url = this.rpOptions.url.endsWith('/') ? this.rpOptions.url.slice(0, -1) : this.rpOptions.url;
        const shard = this.pwConfig.shard;
        const uploadUrl = new URL('/api/result/upload', url);
        // set specific parameter with file content length, to handle s3 presigned url upload
        uploadUrl.searchParams.set('fileContentLength', buffer?.length?.toString() ?? '0');
        const resp = await ctx.put(uploadUrl.href, {
            multipart: {
                file: {
                    name: this.blobName ?? 'blob.zip',
                    mimeType: 'application/zip',
                    buffer: buffer,
                },
                // Passing result details from user, and internal
                ...clearedResDetails,
                ...(shard ? { shardCurrent: shard.current, shardTotal: shard.total } : {}),
                triggerReportGeneration: this.rpOptions.triggerReportGeneration,
                playwrightVersion: version,
            },
        });
        const resultResponse = (await resp.json()).data;
        console.debug('[ReporterPlaywrightReportsServer] blob result uploaded: ', resultResponse);
        // If we are not in a shard, we should trigger report generation with POST request, otherwise report will be generated by the server
        if (this.rpOptions.triggerReportGeneration) {
            let report;
            if (this.pwConfig.shard) {
                report = {
                    reportUrl: `/api/serve/${resultResponse.generatedReport?.reportId}/index.html`,
                };
            }
            else {
                report = await (await ctx.post(`${this.rpOptions.url}/api/report/generate`, {
                    data: {
                        resultsIds: [resultResponse.resultID],
                        ...clearedResDetails,
                        playwrightVersion: version,
                    },
                })).json();
            }
            if (report.reportUrl) {
                console.log(`[ReporterPlaywrightReportsServer] 🎭 HTML Report is available at: ${this.rpOptions.url}${report.reportUrl}`);
            }
        }
    }
}
exports.default = ReporterPlaywrightReportsServer;
